File: frontend/src/types/standardization.ts
/** 来源位置项（对应 SourceLocationItem） */
export interface SourceLocationItem {
  type: string;
  sheet?: string | null;
  cell?: string | null;
  table_index?: number | null;
  row?: number | null;
  col?: number | null;
  page?: number | null;
  extraction_mode?: string | null;
  ocr_confidence?: number | null;
}

export type SourceLocation = Record<string, SourceLocationItem>;

/** 规则命中快照（对应 HitRuleSnapshot） */
export interface HitRuleSnapshot {
  rule_id: string;
  rule_name: string;
  match_content: string;
  match_mode: string;
}

/** 标准化行（对应 StandardizedRowResponse） */
export interface StandardizedRow {
  id: string;
  rawTableId: string;
  supplierFileId: string;
  rowIndex: number;
  product_name: string | null;
  spec_model: string | null;
  unit: string | null;
  quantity: number | null;
  unit_price: number | null;
  total_price: number | null;
  tax_rate: string | null;
  delivery_period: string | null;
  remark: string | null;
  sourceLocation: SourceLocation;
  columnMapping: Record<string, string> | null;
  hitRuleSnapshots: HitRuleSnapshot[] | null;
  confidence: number;
  isManuallyModified: boolean;
  needsReview: boolean;
  taxBasis: string | null;
}

/** 列名映射信息（对应 column-mapping-info API 响应） */
export interface ColumnMappingInfo {
  originalColumn: string;
  targetField: string | null;
  matchedRule: string | null;
  matchMode: string | null;
  status: "confirmed" | "pending" | "unmapped" | "conflict";
}

/** 手工修正响应（对应 FieldModifyResponse） */
export interface FieldModifyResponse {
  success: boolean;
  auditLog: {
    field: string;
    beforeValue: string;
    afterValue: string;
    timestamp: string;
  };
  dirtyStages?: string[];
}

/** 标准字段的键名（9 个标准字段） */
export type StandardFieldKey =
  | "product_name"
  | "spec_model"
  | "unit"
  | "quantity"
  | "unit_price"
  | "total_price"
  | "tax_rate"
  | "delivery_period"
  | "remark";

/** 标准字段显示名 */
export const STANDARD_FIELD_LABELS: Record<StandardFieldKey, string> = {
  product_name: "品名",
  spec_model: "规格型号",
  unit: "单位",
  quantity: "数量",
  unit_price: "单价",
  total_price: "总价",
  tax_rate: "税率",
  delivery_period: "交货期",
  remark: "备注",
};

/** 必填字段 */
export const REQUIRED_FIELDS: StandardFieldKey[] = [
  "product_name",
  "unit_price",
  "quantity",
];

File: backend/models/standardization.py
from __future__ import annotations

from pydantic import BaseModel, Field


class SourceLocationItem(BaseModel):
    """单个字段的来源定位"""

    type: str  # xlsx / docx / pdf / pdf_ocr
    sheet: str | None = None
    cell: str | None = None
    table_index: int | None = None
    row: int | None = None
    col: int | None = None
    page: int | None = None
    extraction_mode: str | None = None  # structure / ocr
    ocr_confidence: float | None = None


# source_location 类型：key 为标准字段名，value 为 SourceLocationItem
SourceLocation = dict[str, SourceLocationItem]


class HitRuleSnapshot(BaseModel):
    """规则命中快照"""

    rule_id: str
    rule_name: str
    match_content: str  # 如 "报价→unit_price"
    match_mode: str  # exact / fuzzy / regex


class StandardizedRowCreate(BaseModel):
    """标准化行写入模型"""

    id: str
    raw_table_id: str
    supplier_file_id: str
    row_index: int
    product_name: str | None = None
    spec_model: str | None = None
    unit: str | None = None
    quantity: float | None = None
    unit_price: float | None = None
    total_price: float | None = None
    tax_rate: str | None = None
    delivery_period: str | None = None
    remark: str | None = None
    source_location: SourceLocation
    column_mapping: dict[str, str] | None = None
    hit_rule_snapshots: list[HitRuleSnapshot] | None = None
    confidence: float = 1.0
    needs_review: bool = False
    tax_basis: str | None = None  # known_inclusive / known_exclusive / unknown


class StandardizedRowResponse(BaseModel):
    """标准化行响应模型"""

    id: str
    raw_table_id: str = Field(..., alias="rawTableId")
    supplier_file_id: str = Field(..., alias="supplierFileId")
    row_index: int = Field(..., alias="rowIndex")
    product_name: str | None = None
    spec_model: str | None = None
    unit: str | None = None
    quantity: float | None = None
    unit_price: float | None = None
    total_price: float | None = None
    tax_rate: str | None = None
    delivery_period: str | None = None
    remark: str | None = None
    source_location: SourceLocation = Field(..., alias="sourceLocation")
    column_mapping: dict[str, str] | None = Field(None, alias="columnMapping")
    hit_rule_snapshots: list[HitRuleSnapshot] | None = Field(
        None, alias="hitRuleSnapshots"
    )
    confidence: float = 1.0
    is_manually_modified: bool = Field(False, alias="isManuallyModified")
    needs_review: bool = Field(False, alias="needsReview")
    tax_basis: str | None = Field(None, alias="taxBasis")

    model_config = {"populate_by_name": True}


class FieldModifyRequest(BaseModel):
    """手工修正请求"""

    field: str
    new_value: str | float | None = Field(..., alias="newValue")

    model_config = {"populate_by_name": True}


class FieldModifyResponse(BaseModel):
    """手工修正响应"""

    success: bool
    audit_log: dict = Field(..., alias="auditLog")
    dirty_stages: list[str] = Field(default_factory=list, alias="dirtyStages")

    model_config = {"populate_by_name": True}


class StandardizeRequest(BaseModel):
    """标准化执行请求（异步）"""

    force: bool = False


class StandardizeTaskResponse(BaseModel):
    """标准化执行响应"""

    task_id: str = Field(..., alias="taskId")

    model_config = {"populate_by_name": True}
